Info banner screen reader improvements

jacortinas opened the pull request:
## Issues
- https://github.com/github/accessibility-audits/issues/5635
- https://github.com/github/accessibility-audits/issues/5639
- https://github.com/github/accessibility-audits/issues/5640

## Description
This PR improves screen reader support for banner content when it appears.

The implementation is trying to follow along with some of the lessons learned for the dotcom implementation https://github.com/github/github/pull/242037.

- Adds `role="alert"` and `aria-atomic="true"` to banner container `div`.
- Dismissal of banners is now dependent on the current keyboard focus leaving the banner element.
- The implementation for banners waits 200ms then manually focuses on the first link or button within the banner.
- Add 3 development menu items for popping up banners for testing.

### Screenshots
![Screenshot 2023-10-18 at 10 24 22 AM](https://github.com/desktop/desktop/assets/171215/2ea616d8-f23d-4754-beab-94c1de550261)

## Release notes

Notes: Improve keyboard and screen reader support for banners

jacortinas committed:
Add dev menu items for showing fake banners

jacortinas committed:
Add alert role to banner component

jacortinas committed:
Try aria-atomic and modifying content

jacortinas committed:
Convert links to button for autofocus

jacortinas committed:
Remove auto timeout and content suffix

jacortinas committed:
Revert test buttons to LinkButton components

jacortinas committed:
Add fake cherry pick banner dev menu item

jacortinas committed:
Manually move focus to close or the first anchor

tidy-dev commented on app/src/ui/banners/open-thank-you-card.tsx:
```diff
@@ -31,8 +31,7 @@ export class OpenThankYouCard extends React.Component<
           />
-          or{' '}
-          <LinkButton onClick={this.onThrowCardAway}>Throw It Away</LinkButton>{' '}
+          or <button onClick={this.onThrowCardAway}>Throw It Away</button>{' '}
```
I think we should probably leave these as links unless we are considering changing all the links in the banners to buttons.

jacortinas replied:
I missed this one, oops. Fixed now.

tidy-dev commented on app/src/ui/banners/banner.tsx:
```diff
@@ -4,49 +4,64 @@ import * as OcticonSymbol from '../octicons/octicons.generated'
 
 interface IBannerProps {
   readonly id?: string
-  readonly timeout?: number
```
Sorry.. I maybe have been confusing when talking about this in the team sync, but I do not think we can completely do away with this as right now our current users have muscle memory around these automatically disappearing.

If we are going to remove them, we also need to add in the logic that dismisses them when they lose focus.

Otherwise, I think it is just prudent to elongate the timer on the banners that have undo buttons.

tidy-dev replied:
I tinkered with this a bit after leaving this comment. Here is a branch with the focus management logic I am talking about. https://github.com/desktop/desktop/compare/jc-reorder-alert...timer-on-focus-out?expand=1

jacortinas replied:
@tidy-dev Thank you for the branch 🙇 , I've integrated those changes into this PR.

tidy-dev commented on app/src/ui/banners/banner.tsx:
```diff
@@ -4,49 +4,64 @@ import * as OcticonSymbol from '../octicons/octicons.generated'
-    }
+  public componentDidMount(): void {
+    this.focusTimeoutId = window.setTimeout(() => {
+      if (this.closeButton.current) {
+        this.closeButton.current.focus()
+      } else {
+        this.contents.current?.querySelector('a')?.focus()
```
I believe we should have this inverted in that if there is a link in the contents, that gets focused and if not then focus the dismiss button. This is the pattern we follow on our dialogs as well and has roots in the accessibility concept of "focus the first interactive element". 

Other thoughts are to pull this into a method like in the `dialog.tsx` and use the `sleep` helper method to make this code a little easier to read and could add some documentation - especially explaning the need for the timeout. Something like: 
```
  private async focusFirstSuitableElement() {
     // Timeout is required because flows such as rebase are in dialogs and have focus management that sets the focus to the app body after this has run.
    await sleep(200)
    
     const link = this.contents.current?.querySelector('a')
     if (link) {
      link.focus()
      return
    }
    
    this.closeButton.current?.focus()
  }
```

MrDoctorAlmaty replied:
```suggestion
      if (this.closeButton.current) {
        this.closeButton.current.focus()
      } else {
        this.contents.current?.querySelector('a')?.focus()
```

jacortinas replied:
@tidy-dev I've looked at the implementation of `sleep`, and I think the only issue might be that there is no simple way to hold onto the ID's that come from the `window.setTimeout` calls. Right now AFAICT we make sure to clean up our timeouts to avoid calling `onDismiss` or moving focus to non-existent elements if the component is unmounted.

tidy-dev reviewed (changes requested):
@jacortinas This is looking great. I have a few comments, and I found one behavior issue that I am not sure how to solve at the moment. 

Issue repro steps:
1. added two no-conflicting commits
2. reordered.
3. Used keyboard to click "undo"
4. The voice over correctly reads the "Undone" message.. but the focus does not actually change to the button. 
![Showing focus still on phantom undo link](https://github.com/desktop/desktop/assets/75402236/b4712e01-309a-47dd-b9c0-b42b68ff670d)

I believe the reason is that the react `<Banner>` is still rendered with different children so the `componentDidMount` does not fire again.

jacortinas committed:
Missed a button tag

jacortinas committed:
Add focus based dismissal

jacortinas committed:
Remove overflow hidden styles for focus outline

Without removing some overflow styles, the focus outline for elements inside of the banner is cut off and difficult to see.

jacortinas committed:
Merge branch 'development' into jc-reorder-alert

jacortinas commented:
> @jacortinas This is looking great. I have a few comments, and I found one behavior issue that I am not sure how to solve at the moment.
> 
> Issue repro steps:
> 
> 1. added two no-conflicting commits
> 2. reordered.
> 3. Used keyboard to click "undo"
> 4. The voice over correctly reads the "Undone" message.. but the focus does not actually change to the button.
>    ![Showing focus still on phantom undo link](https://user-images.githubusercontent.com/75402236/276657940-b4712e01-309a-47dd-b9c0-b42b68ff670d.png)
> 
> I believe the reason is that the react `<Banner>` is still rendered with different children so the `componentDidMount` does not fire again.

@tidy-dev I see the problem you're mentioning, I agree that it's a blocker. I'm working on a solution. :+1:

jacortinas committed:
Show a fake undone banner when undone

jacortinas committed:
Cleaner banner functions

jacortinas committed:
Move each type of banner into different components

jacortinas commented:
@tidy-dev ~~Recent commits fix the issue with the banner components having the wrong focus after clicking "undo". I solved it by created individual components for the last couple of `SuccessBanner` uses in `render-banner.jsx`. That way when the banner is changed from one type to another, the child `SucessBanner` component is truly unmounted and remounted.~~

Update: Nevermind, turns out that the different banner components just needed to have a `key` prop added for each one. Problem is fixed.

jacortinas commented:
```diff
diff --git a/app/src/ui/app.tsx b/app/src/ui/app.tsx
index 5139f443cb..ed1e12d549 100644
--- a/app/src/ui/app.tsx
+++ b/app/src/ui/app.tsx
@@ -3175,7 +3175,11 @@ export class App extends React.Component<IAppProps, IAppState> {
     return (
       <TransitionGroup>
         {banner && (
-          <CSSTransition classNames="banner" timeout={bannerTransitionTimeout}>
+          <CSSTransition
+            classNames="banner"
+            timeout={bannerTransitionTimeout}
+            onExited={this.onBannerTransitionExit}
+          >
             {banner}
           </CSSTransition>
         )}
@@ -3183,6 +3187,19 @@ export class App extends React.Component<IAppProps, IAppState> {
     )
   }
 
+  private onBannerTransitionExit = () => {
+    if (
+      document
+        .querySelector('#desktop-app-contents .banner')
+        ?.contains(document.activeElement)
+    ) {
+      const branchButton = document.querySelector(
+        '#desktop-app-toolbar .branch-button button'
+      ) as HTMLButtonElement
+      branchButton?.focus()
+    }
+  }
+
   private renderUpdateBanner() {
     return (
       <UpdateAvailable
```

@tidy-dev As mentioned in [this comment](https://github.com/github/accessibility-audits/issues/5635#issuecomment-1696391935), here is a potential solution for moving focus after the dismissal of a banner. There is one interesting caveat that we don't want to steal focus if the user actively starts moving it around and trigger the auto-dismissal so there is a check to make sure the currently active element is not a child of a banner element. I think I might be able to make it slightly cleaner if I add some ID's to specific components I'm targeting but it's the same basic implementation. Thoughts?

jacortinas committed:
Revert "Move each type of banner into different components"

This reverts commit fce2052af4e02e8f6c88e0c2ce0584d4e83a1c82.

jacortinas committed:
Use distinct key props instead

jacortinas committed:
Ooops this was 15000

tidy-dev commented on app/src/ui/banners/banner.tsx:
```diff
@@ -10,43 +10,99 @@ interface IBannerProps {
+  }
+
+  private focusOnFirstSuitableElement = () => {
+    const target =
+      this.banner.current?.querySelector('a') ||
+      this.banner.current?.querySelector('button')
+    target?.focus()
```
I love how concise this is. 💖

tidy-dev reviewed (approved):
✨  Thank you for all your iterations on this! It works beautifully!

Re-cap of our discussion in slack on manual focus after banner close is to not include it as part of these fixes. Especially given concerns of accident focus grabbing that it could lead to.
